### PR TITLE
enhancement: overhaul refresh token function

### LINKED ISSUE #29: Is it safe to trust old tokens?
If it is, we can save 1 db hit. And the main purpose of authomatic stores will be to tell us whether refresh tokens have been invalidated.

### PR DESCRIPTION
**Pull request by Amri91**
Fixes #29

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -39,10 +37,18 @@ const Secret = secret => {
...
 const ExpiresAt = t.Integer;
 const Algorithm = t.enums.of(['HS256', 'HS384', 'HS512', 'RS256'], 'Algorithm');
-const UserId = t.Any;
-const RefreshToken = t.String;
-const Token = t.String;
+const UserId = t.String;
 const Prolong = t.Boolean;
+const AccessTokenType = 'AT';
+const RefreshTokenType = 'RT';
+const RefreshTokenSignOptions = {
```
Comment (Amri91) [inline index.js]: Should nbt be added?

Comment (Amri91) [reply index.js]: Pros: Lower refresh requests, prevents a kind of DOS?
Cons: Harder for clients to use

File: index.js [hunk 2]
```diff
@@ -191,71 +224,101 @@ module.exports = function Authomatic({
   SignOptions(defaultSignOptions);
   VerifyOptions(defaultVerifyOptions);
 
+  const Token = type => token => {
+    const decodedContent = jwt.decode(token);
+    if(decodedContent && decodedContent.t === type) {
```
[the remark this answers is not in the document]
Comment (Amri91) [reply index.js]: I am gonna try adding it to the aud claim.

File: index.js [hunk 3]
```diff
@@ -26,6 +27,10 @@ const JWT = t.interface({
   decode: t.Function,
 }, 'JWT');
 
+const uniqueAdd = element => ifElse(contains(element), identity, append(element));
+const audLens = lensProp('aud');
+const audienceLens = lensProp('audience');
```
Comment (Avaq) [inline index.js]: You're bringing a gun to a knife fight

File: index.js [hunk 4]
```diff
@@ -191,71 +229,93 @@ module.exports = function Authomatic({
   SignOptions(defaultSignOptions);
   VerifyOptions(defaultVerifyOptions);
 
+  const tokensMatch = (ATContent, RTContent) => {
+    if (RTContent.accessTokenJTI !== ATContent.jti) {
+      throw tokensMismatch;
+    }
+    return true;
```
Comment (Avaq) [inline index.js]: This return value is never used.

**Comment by Avaq** [comment]
> Can you explain what you changed?

**Comment by Amri91** [comment]
Amended, thank you all for the comments.

**Comment by Amri91** [comment]
adding claim t

**Comment by Amri91** [comment]
Added a new commit to make review easier, I will rebase and point to authomatic-redis at the end.

File: jsdoc2md/readme.hbs [hunk 5]
```diff
@@ -1,12 +1,12 @@
...
 [![Maintainability](https://api.codeclimate.com/v1/badges/314b595549aca68c5c6c/maintainability)](https://codeclimate.com/github/wearereasonablepeople/authomatic/maintainability)
-[![Test Coverage](https://api.codeclimate.com/v1/badges/314b595549aca68c5c6c/test_coverage)](https://codeclimate.com/github/wearereasonablepeople/authomatic/test_coverage)
+[![Coverage Status](https://coveralls.io/repos/github/wearereasonablepeople/authomatic/badge.svg?branch=master)](https://coveralls.io/github/wearereasonablepeople/authomatic?branch=master)
 [![dependencies Status](https://david-dm.org/wearereasonablepeople/authomatic/status.svg)](https://david-dm.org/wearereasonablepeople/authomatic)
 [![devDependencies Status](https://david-dm.org/awearereasonablepeople/authomatic/dev-status.svg)](https://david-dm.org/wearereasonablepeople/authomatic?type=dev)
 
 ## Description
-An opinionated JWT library with sensible defaults that supports refresh and access tokens.
+An authentication library that uses JWT for access and refresh tokens with sensible defaults.
```
Comment (Avaq) [inline jsdoc2md/readme.hbs]: Also changed in package.json, github repo description?

**Comment by Amri91** [comment]
Rebased

**Pull request merged**